euphonic/design/03_design_document_v0.3.md

```diff
@@ -0,0 +1,385 @@
+
+|Attribute|Type|
+|---------|----|
+|x bin edges|(n_xbins + 1), array of floats|
+|x bin units|string|
+|x tick labels|list of strings|
+|x tick locations|array of ints|
+|plot data|(n_xbins,) array of floats|
+
+|Method|Input required|Output|
+|------|--------------|------|
+|Plot|x broadening|Matplotlib Figure|
+|Serialise|Filename|File
+
+## Python data object types
+As can be seen above, the Force Constants and Phonon Data objects are the only
+ones that do complex calculations. The other objects generally just contain
+calculated data, and maybe do some simple calculations. For this reason Force
+Constants and Phonon Data will be Python `classes` and all the others will be
+Python [`dataclasses`](https://docs.python.org/3/library/dataclasses.html).
+
+`dataclasses` have been chosen as they require less boilerplate code than
+regular classes, yet they still allow methods to be written and have a type so
+their purpose and content is obvious to the user. `dataclasses` were only
+introduced in Python `3.7` (and backported to `3.6`), but these versions are
+supported by both Matlab and Mantid so this is an acceptable dependency.
+
+### Alternatives
+[**NamedTuple**](https://docs.python.org/3/library/typing.html#typing.NamedTuple) - 
+Typed `typing.NamedTuple` or untyped `collections.namedtuple` were considered
+as a possible alternative because they have attribute access, can contain all
+the required data and could be subclassed to add methods. However, they were not
+chosen as they require more boilerplate to subclass, are not as flexible, and
+although immutability could be seen as an advantage as it guarantees the data
+will not change, this could have unforseen restrictions/inefficiencies in the
+future. In addition many features of `typing.NamedTuple` were not added until
+`3.6` so version restrictions would be similar
+
+**Class** - Regular Python classes were a potential option as they could satisfy
+all requirements of containing data and having methods, but might be
+overkill for some very simple data objects such as DOS, contain lots of
+boilerplate code, and therefore be more difficult to maintain.
+
+**Dictionary** - Dictionaries could contain all the required data, but as they
+are not a class won't have an obvious type so the user would have to inspect the
+keys to see what it contains. In addition, any functions for this data would
+have to lie elsewhere in the code, and dictionary access (`obj['a']`) would
+clash syntactically with attribute access (`obj.a`) elsewhere.
+
+**Numpy Array** - Very simple, but wouldn't be able to bundle associated data
+together and would require the user to keep track of this. For serialisation,
+plotting, further calculations the user would have to pass all pieces of
+required data as arguments which is messy and likely to cause errors.
+
+## Units
+Units in Euphonic will be handled by the [`Pint`](https://pint.readthedocs.io)
```

Comment:
I've been playing with dataclasses/classes and pint units and I think if a user is instantiating an object directly, they should wrap any data with units as a pint `Quantity` object (which would be made clear in the docstring). What is less clear is what we should do for intput/output, from files or other data structures such as dictionaries. One option is to output the data as a numpy array with a string specifying the units e.g. `cell_vectors=np.ndarray([...])` and `cell_vectors_unit='angstrom'`, however there are also options for converting `Quantity` objects to a tuple of magnitude and unit for file output, which I'm also looking into, but this could potentially be inefficient.

I've done a quick prototype of the `Structure` object mentioned above which shows how this could work, which I'll upload and I'll generally update the doc to make this clearer. On another note my prototype shows that classes might actually be better than dataclasses for dealing with this unit stuff...